File: apps/backend/internal/model/community/community.go
package community

import (
	"time"

	"github.com/google/uuid"
	"github.com/sarbojitrana/nexus/internal/model"
	"github.com/sarbojitrana/nexus/internal/model/post"
)

type PostPermission string

const (
	AllPost        PostPermission = "all"
	ModeratorsOnly PostPermission = "moderator"
)

type CommunityRole string

const (
	MemberRole CommunityRole = "member"
	AdminRole  CommunityRole = "admin"
	CommonRole CommunityRole = "all"
)

type CommunityReportStatus string

const (
	ReportPending   CommunityReportStatus = "pending"
	ReportResolved  CommunityReportStatus = "resolved"
	ReportDismissed CommunityReportStatus = "dismissed"
)

type Community struct {
	model.Base
	AdminID      uuid.UUID `json:"adminId" db:"admin_id"`
	Name         string    `json:"name" db:"name"`
	Slug         string    `json:"slug" db:"slug"`
	Description  *string   `json:"description" db:"description"`
	AvatarKey    *string   `json:"avatarKey" db:"avatar_key"`
	BannerKey    *string   `json:"bannerKey" db:"banner_key"`
	MembersCount int       `json:"membersCount" db:"members_count"`
	PostsCount   int       `json:"postsCount" db:"posts_count"`
}

type CommunityMember struct {
	UserID      uuid.UUID     `json:"userId" db:"user_id"`
	CommunityID uuid.UUID     `json:"communityId" db:"community_id"`
	Role        CommunityRole `json:"role" db:"role"`
	JoinedAt    time.Time     `json:"joinedAt" db:"joined_at"`
	model.BaseWithUpdatedAt
}
type MiniCommunity struct {
	CommnunityID uuid.UUID `json:"communityId" db:"community_id"`
	Name         string    `json:"communityName" db:"name"`
	AvatarKey    string    `json:"communityAvatarKey" db:"avatar_key"`
}
type CommnunitySummary struct {
	MiniCommunity
	MembersCount int `json:"membersCount" db:"members_count"`
	PostsCount   int `json:"postsCount" db:"posts_count"`
}

type ViewCommunityPost struct {
	post.Post
	PostMedia []post.PostMedia `json:"postMedia" db:"post_media"`
	Community *MiniCommunity   `json:"miniCommunity" db:"mini_community"`
}

type CommnunityFollow struct {
	model.BaseWithId
	model.BaseWithCreatedAt
	FollowerID  uuid.UUID `json:"followerId" db:"follower_id"`
	CommunityID uuid.UUID `json:"communityId" db:"community_id"`
}

type CommunityReport struct {
	model.Base
	ReporterID   uuid.UUID             `json:"reporterId" db:"reporter_id"`
	CommnunityID uuid.UUID             `json:"communityId" db:"community_id"`
	PostID       uuid.UUID             `json:"postId" db:"post_id"`
	Reason       string                `json:"reason" db:"reason"`
	Status       CommunityReportStatus `json:"status" db:"status"`
}

type MiniCommunityUser struct {
	UserID    uuid.UUID     `json:"userId" db:"user_id"`
	AvatarKey string        `json:"avatarKey" db:"avatar_key"`
	Name      string        `json:"name" db:"name"`
	JoinedAt  time.Time     `json:"joinedAt" db:"joined_at"`
	Role      CommunityRole `json:"role" db:"role"`
}

File: packages/zod/src/community.ts
import { z } from "zod";
import { ZOrder, schemaWithCursorPagination } from "@/utils.js";

export const ZCommunityRole = z.enum(["member", "moderator", "admin"]);
export const ZPostPermission = z.enum(["all", "moderator"]);
export const ZCommunityReportStatus = z.enum([
  "pending",
  "resolved",
  "dismissed",
]);

export const ZCommunity = z.object({
  id: z.string().uuid(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  adminId: z.string(),
  name: z.string(),
  slug: z.string(),
  description: z.string().nullable(),
  avatarKey: z.string().nullable(),
  bannerKey: z.string().nullable(),
  membersCount: z.number(),
  postsCount: z.number(),
  canPost: ZPostPermission.nullable(),
});

export const ZCommunityResponse = ZCommunity.extend({
  viewerRole: ZCommunityRole.nullable(),
});
export type Community = z.infer<typeof ZCommunity>;
export type CommunityResponse = z.infer<typeof ZCommunityResponse>;

export const ZCommunityMember = z.object({
  userId: z.string(),
  communityId: z.string().uuid(),
  role: ZCommunityRole,
  joinedAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const ZMiniCommunity = z.object({
  communityId: z.string().uuid(),
  slug: z.string(),
  communityName: z.string(),
  communityAvatarKey: z.string().nullable(),
  membersCount: z.number(),
  postsCount: z.number(),
  createdAt: z.string().datetime(),
});
export type MiniCommunity = z.infer<typeof ZMiniCommunity>;

export const ZMiniCommunityUser = z.object({
  userId: z.string(),
  avatarKey: z.string().nullable(),
  name: z.string(),
  joinedAt: z.string().datetime(),
  role: ZCommunityRole,
});
export type MiniCommunityUser = z.infer<typeof ZMiniCommunityUser>;
export type CommunityRole = z.infer<typeof ZCommunityRole>;

export const ZCommunityReport = z.object({
  id: z.string().uuid(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
  reporterId: z.string(),
  communityId: z.string().uuid(),
  postId: z.string().uuid(),
  reason: z.string(),
  status: ZCommunityReportStatus,
});
export type CommunityReport = z.infer<typeof ZCommunityReport>;

export const ZBannedFromCommunityUser = z.object({
  communityId: z.string().uuid(),
  userId: z.string(),
  createdAt: z.string().datetime(),
});

export const ZCreateCommunityPayload = z.object({
  name: z.string().max(50),
  slug: z.string().max(50),
  description: z.string().max(1000).nullable().optional(),
  avatarKey: z.string().nullable().optional(),
  bannerKey: z.string().nullable().optional(),
});

export const ZUpdateCommunitySettingsPayload = z.object({
  name: z.string().max(50).nullable().optional(),
  slug: z.string().max(50).nullable().optional(),
  description: z.string().max(1000).nullable().optional(),
  avatarKey: z.string().nullable().optional(),
  bannerKey: z.string().nullable().optional(),
});

export const ZChangeMemberRoleInCommunityPayload = z.object({
  newRole: ZCommunityRole,
});

export const ZGetCommunitiesQuery = z.object({
  cursorSortValue: z.string().optional(),
  cursorCreatedAt: z.string().datetime().optional(),
  sort: z.enum(["created_at", "members_count"]).optional(),
  order: ZOrder.optional(),
  name: z.string().optional(),
});

export const ZGetCommunityMembersQuery = z.object({
  cursorSortValue: z.string().optional(),
  cursorCreatedAt: z.string().datetime().optional(),
  order: ZOrder.optional(),
  role: z.enum(["all", "moderator", "member"]).optional(),
});

export const ZReportCommunityPostPayload = z.object({
  postId: z.string().uuid(),
  reason: z.string().max(1000),
});

export const ZResolveCommunityPostReportPayload = z.object({
  updatedStatus: z.enum(["resolved", "dismissed"]),
});

export const ZBanCommunityMemberPayload = z.object({
  userIdToBan: z.string(),
});

export const ZGetCommunityReportsQuery = z.object({
  status: ZCommunityReportStatus.optional(),
  reportedDateStart: z.string().optional(),
  reportedDateEnd: z.string().optional(),
  cursorCreatedAt: z.string().datetime().optional(),
});

export const ZGetCommunitiesResponse = schemaWithCursorPagination(ZMiniCommunity);
export const ZGetCommunityMembersResponse =
  schemaWithCursorPagination(ZMiniCommunityUser);
export const ZGetCommunityReportsResponse =
  schemaWithCursorPagination(ZCommunityReport);
